retrieve-logs/fetch_missing_matomo_requests.py

```diff
@@ -64,47 +110,46 @@ def run_query(start_timestamp, end_timestamp):
+    count_written = 0
     with open(output_filename, 'a+') as f:
         for message in response['results']:
+            if len(message) >= MAX_REQUESTS:
+                log_too_many_requests_and_exit(period_start, period_end)
             for message in message:
                 if message['field'] == '@message':
                     f.write(message['value'] + '\n')
+                    count_written += 1
                     break
+    if count_written > 0:
+        get_logger().debug(
+                f'Wrote {count_written} requests to file {output_filename}'
+                + f' from within the period {period_start} to {period_end}')
+    return count_written
 
 
 if __name__ == '__main__':
     validate_environment_variables()
 
     client = boto3.client('logs')
 
-    start_date = get_start_date()
+    start_datetime = get_start_datetime()
     num_of_days = get_number_of_days()
-    end_date = start_date + timedelta(days=num_of_days) + timedelta(microseconds=-1)
-
-    OUTPUT_FILENAME = start_date.strftime('%Y%m%d') + '_' + end_date.strftime('%Y%m%d') + '_matomo_requests.json'
-    if os.path.exists(OUTPUT_FILENAME):
-        os.remove(OUTPUT_FILENAME)
-
-    for days in range(0, get_number_of_days()):
-        current_date = start_date + timedelta(days=(days))
-        end_date = current_date + timedelta(days=1, microseconds=-1)
-
-        start_timestamp = current_date.replace(tzinfo=timezone.utc).timestamp()
-        end_timestamp = end_date.replace(tzinfo=timezone.utc).timestamp()
-
-        duration = (end_date - current_date).total_seconds()
-        offset = 60 * 5
-        num_of_iterations = int(duration / offset)
-
-        for i in range(num_of_iterations):
-            response = run_query(datetime.utcfromtimestamp(start_timestamp),
-                                 (datetime.utcfromtimestamp((start_timestamp + offset)) + timedelta(microseconds=-1)))
-            response = wait_for_the_query_to_complete(response)
-            write_requests_to_a_file(response, start_date, end_date, OUTPUT_FILENAME)
-            start_timestamp = start_timestamp + offset
-        if Decimal(duration) / Decimal(offset) % Decimal(1) != Decimal(0):
-            response = run_query(datetime.utcfromtimestamp(start_timestamp),
-                                 (datetime.utcfromtimestamp((start_timestamp + offset)) + timedelta(microseconds=-1)))
-            response = wait_for_the_query_to_complete(response)
-            write_requests_to_a_file(response, start_date, end_date, OUTPUT_FILENAME)
+    period_width = get_period_width()
+    end_datetime = start_datetime + timedelta(days=num_of_days, microseconds=-1)
+
+    output_filename = get_output_filename(start_datetime, end_datetime)
+    if os.path.exists(output_filename):
+        os.remove(output_filename)
+
+    period_start = datetime.utcfromtimestamp(start_datetime.replace(tzinfo=timezone.utc).timestamp())
+    total_written = 0
+    while period_start < end_datetime:
```

Comment:
Should this be `<=` given you done the `microseconds=-1`?

Reply:
Yep, thanks for spotting. Have fixed in a new commit.